### ISSUE TITLE
[Bug] VSTestV2: collect code coverage throws errors after updating to VS Enterprise 16.9.1 and private agent to version 2.183.1

### ISSUE THREAD

**Issue by bwstaal** [microsoft/azure-pipelines-tasks#14594]
**Question, Bug, or Feature?**  
*Type*: Bug

**Task Name**: VSTestV2

## Environment
- Azure Pipelines

- _Agent:_ private
    
_OS:_ Windows Server 2012 R2
_Agent Version:_ 2.183.1

## Issue Description

After updating Visual Studio to version 16.9.1 and all the build agents to version 2.183.1 on the CI server at my employer, collecting code coverage for a .Net 4.6.1 solution doesn't work anymore. Below are the settings of the VSTestV2 task and the error message that is being thrown. Updating the NuGet packages "MSTest.TestAdapter" and "MSTest.TestFramework" to version 2.2.2, including installing/updating any related NuGet packages unfortunately didn't help. I tried this, because according to the [release notes](https://github.com/microsoft/testfx-docs/blob/master/docs/releases.md), support for TestPlaform v16.9.1 was added.

**VSTest V2 'Test Assemblies' task settings:**
#Your build pipeline references an undefined variable named ‘BuildPlatform’. Create or edit the build pipeline for this YAML file, define the variable on the Variables tab. See https://go.microsoft.com/fwlink/?linkid=865972
#Your build pipeline references an undefined variable named ‘BuildConfiguration’. Create or edit the build pipeline for this YAML file, define the variable on the Variables tab. See https://go.microsoft.com/fwlink/?linkid=865972

```
steps:
- task: VSTest@2
  displayName: 'Test Assemblies'
  inputs:
    testAssemblyVer2: |
     **\*test*.dll
     !**\obj\**
    searchFolder: '$(Build.BinariesDirectory)'
    runOnlyImpactedTests: true
    runAllTestsAfterXBuilds: 10
    runInParallel: true
    codeCoverageEnabled: true
    platform: '$(BuildPlatform)'
    configuration: '$(BuildConfiguration)'
    diagnosticsEnabled: true
```
_BuildConfiguration:_ release
_BuildPlatform:_ any cpu

**Error message**

##[error]Data collector 'Code Coverage' message: Data collector 'Code Coverage' failed to provide initialization information. Error: System.ArgumentNullException: Value cannot be null.
##[error]Parameter name: first
##[error]   at System.Linq.Enumerable.Concat[TSource](IEnumerable`1 first, IEnumerable`1 second)
##[error]   at Microsoft.VisualStudio.Coverage.DynamicCoverageDataCollector.GetEnvironmentVariables()
##[error]   at Microsoft.VisualStudio.TestPlatform.Common.DataCollector.DataCollectorInformation.SetTestExecutionEnvironmentVariables()
##[error]   at Microsoft.VisualStudio.TestPlatform.Common.DataCollector.DataCollectionManager.GetEnvironmentVariables(Boolean& unloadedAnyCollector).
##[error]Data collector 'Test Impact' message: No diagnostics or data will be collected because the diagnostic data adapters did not initialize..
Blame: Attaching crash dump utility to process testhost.net461.x86 (3396).
Blame: Attaching crash dump utility to process testhost.net461.x86 (1340).
##[error]Testhost process exited with error: . Please check the diagnostic logs for more information.
##[error]Testhost process exited with error: . Please check the diagnostic logs for more information.
Blame: Attaching crash dump utility to process testhost.net461.x86 (4268).
##[error]Testhost process exited with error: . Please check the diagnostic logs for more information.
Blame: Attaching crash dump utility to process testhost.net461.x86 (5752).
##[error]Testhost process exited with error: . Please check the diagnostic logs for more information.
2.7669
6.5914
2.8132
... and so forth

**Comment by anshii03** [comment]
@bwstaal  Can you please set System.debug = true , and share full logs with us to investigate ?

**Comment by bwstaal** [comment]
@anshii03 Yes, sure! But due to all the sensitive data in them, I rather don't want to share them publicly here. So what's the best way to provide them to you?

**Comment by anshii03** [comment]
@bwstaal  Please mail to devops_tools@microsoft.com

**Comment by bwstaal** [comment]
@anshii03 Alright, I will collect and email them as soon I have some time available to do so.

**Comment by Craigoss** [comment]
We are also getting this same error since updating to 16.9.4. Is there an update on this issue?

**Comment by ganesp** [comment]
@bwstaal , @Craigoss 
Can you please try the following workaround?
Use Visual Studio Test Platform Tools Installer task in your pipeline, insert it before Visual Studio Test task (VSTestV2) and specify the VSTestV2 task to use the test platform installed by the tools installer in previous task.

Please go through this document for more details:
https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/tool/vstest-platform-tool-installer?view=azure-devops

**Comment by Craigoss** [comment]
@ganesp 
Changes worked for us, thanks for the feedback.

**Comment by bwstaal** [comment]
@ganesp I've tried to add the Visual Studio Test Platform Tools Installer task and modified the Visual Studio Test task (VSTestV2) to use the test platform version that was installed by the Visual Studio Test Platform Tools Installer task. Unfortunately for me, collecting code coverage still doesn't work. FYI: at the moment VS2019 16.9.6 and Azure DevOps pipelines agent v2.186.1 are at the moment installed on our CI Azure VM.

I've put the system.debug variable to true, to collect some logs. I've sent those to devops_tools@microsoft.com as asked by @anshii03

Hopefully those logs will help you to investigate the problem further.

**Comment by ganesp** [comment]
@bwstaal Sorry for the delay here. 
From the logs you shared, I see that the latest stable version was 16.9.4.

Can you please re-try as I see new latest stable version available is 16.10.0
https://www.nuget.org/packages/Microsoft.TestPlatform

If the problem still repros then please do share the logs.

**Comment by bwstaal** [comment]
@ganesp I have some good news! The issue is solved!

After upgrading Visual Studio to version 16.10.1, the Azure DevOps pipelines agent to v2.187.2, and updating the NuGet packages "MSTest.TestAdapter" and "MSTest.TestFramework" to their latest versions, I triggered a build with collecting code coverage enabled. Before I did that, I configured the pipeline, to let the Visual Studio Test Platform Tools Installer task, install the latest test platform version (16.10.0) and the Visual Studio Test task (VSTestV2) to use that version. That build completed successfully. Then I triggered another build, but this time with the Visual Studio Test task (VSTestV2) configured to use the test platform version that's installed with the latest Visual Studio version. That build also completed successfully.

It would be nice, if you can share what was causing this problem in Visual Studio version 16.9.x

**Comment by jakubch1** [comment]
@bwstaal thanks for reporting this. It's fixed only in 16.10. We received this issue https://github.com/microsoft/vstest/issues/2822. It was regression in 16.9 but as there was no a lot of reports we decided to not service 16.9.
